### DISCUSSION TITLE
用MAA刷萨米肉鸽隐藏坍缩范式的简易实现方法

### DISCUSSION THREAD

**Discussion by Alan-Charred** [MaaAssistantArknights/MaaAssistantArknights#9040] [Ideas]
## 前情提要

集成战略#4「探索者的银凇止境」中存在着两组隐藏坍缩范式：**目空一些｜睁眼瞎**，以及 **图像损坏｜一抹黑**。详情参考[PRTS.WIKI相关页面](https://prts.wiki/w/%E6%8E%A2%E7%B4%A2%E8%80%85%E7%9A%84%E9%93%B6%E5%87%87%E6%AD%A2%E5%A2%83/%E5%86%AC%E5%A4%9C%E5%B1%95%E8%A7%88%E9%A6%86)。
其极低的出现概率对于我这样无法投入大量时间刷坍缩，又有着全收集强迫症的玩家非常不友好。
好消息是我们有MAA，只要进行简单的操作就可以让它为我们代劳。

## 任务流程

MAA的任务流程主要由resource/tasks.json文件定义。详情参考[MAA任务流程协议](https://maa.plus/docs/%E5%8D%8F%E8%AE%AE%E6%96%87%E6%A1%A3/%E4%BB%BB%E5%8A%A1%E6%B5%81%E7%A8%8B%E5%8D%8F%E8%AE%AE.html)。
简单地分析一下`resource/tasks.json`的内容并与`resource/template`目录下的图片名称对照可以发现，多个以`Sami@Roguelike@Stage`开头的任务代表着点击事件节点：

### 作战

- `Sami@Roguelike@StageCombatDps`
- `Sami@Roguelike@StageCombatDpsAI6`
- `Sami@Roguelike@StageVerticalCombatDps`
- `Sami@Roguelike@StageVerticalCombatDpsAI6`

### 紧急作战

- `Sami@Roguelike@StageEmergencyDps`
- `Sami@Roguelike@StageEmergencyDpsAI6`
- `Sami@Roguelike@StageVerticalEmergencyDps`
- `Sami@Roguelike@StageVerticalEmergencyDpsAI6`

### 险路恶敌

- `Sami@Roguelike@StageDreadfulFoe`
- `Sami@Roguelike@StageDreadfulFoe-5`

### 不期而遇

- `Sami@Roguelike@StageEncounter`
- `Sami@Roguelike@StageEncounterAI6`
- `Sami@Roguelike@StageVerticalEncounter`
- `Sami@Roguelike@StageVerticalEncounterAI6`

### 安全的角落

- `Sami@Roguelike@StageSafeHouse`
- `Sami@Roguelike@StageSafeHouseAI6`
- `Sami@Roguelike@StageVerticalSafeHouse`
- `Sami@Roguelike@StageVerticalSafeHouseAI6`

### 兴致盎然

- `Sami@Roguelike@StageGambling`
- `Sami@Roguelike@StageGamblingAI6`
- `Sami@Roguelike@StageVerticalGambling`
- `Sami@Roguelike@StageVerticalGamblingAI6`

### 命运所指

- `Sami@Roguelike@StageProphecy`
- `Sami@Roguelike@StageProphecyAI6`
- `Sami@Roguelike@StageVerticalProphecy`
- `Sami@Roguelike@StageVerticalProphecyAI6`

### 得偿所愿

- `Sami@Roguelike@StageBoons`
- `Sami@Roguelike@StageBoonsAI6`
- `Sami@Roguelike@StageVerticalBoons`
- `Sami@Roguelike@StageVerticalBoonsAI6`

### 诡意行商

- `Sami@Roguelike@StageTrader`
- `Sami@Roguelike@StageTraderAI6`
- `Sami@Roguelike@StageVerticalTrader`
- `Sami@Roguelike@StageVerticalTraderAI6`

### 失与得

- `Sami@Roguelike@StageWindAndRain`
- `Sami@Roguelike@StageWindAndRainAI6`
- `Sami@Roguelike@StageVerticalWindAndRain`
- `Sami@Roguelike@StageVerticalWindAndRainAI6`

### 先行一步

- `Sami@Roguelike@StageEmergencyTransportation`
- `Sami@Roguelike@StageEmergencyTransportationAI6`
- `Sami@Roguelike@StageVerticalEmergencyTransportation`
- `Sami@Roguelike@StageVerticalEmergencyTransportationAI6`

### 树篱之途

- `Sami@Roguelike@StageBoskyPassage`
- `Sami@Roguelike@StageBoskyPassageAI6`
- `Sami@Roguelike@StageVerticalBoskyPassage`
- `Sami@Roguelike@StageVerticalBoskyPassageAI6`

### 模糊的预感

- `Sami@Roguelike@StageMysteriousPresage`
- `Sami@Roguelike@StageFerociousPresage`

在执行这些任务后，MAA会点击相应的节点，入下图所示。
![screenshot_20240509_113631](https://github.com/MaaAssistantArknights/MaaAssistantArknights/assets/44151844/e0880dae-9c40-495f-99aa-e07572348b26)
而点击**出发前往**则是下一项任务（此处应当是`Sami@Roguelike@StageCombatDpsEnter`）。
而此时我们是可以查看坍缩范式的，入下图所示。
![screenshot_20240509_113944](https://github.com/MaaAssistantArknights/MaaAssistantArknights/assets/44151844/c5e85300-2feb-404b-b9e0-19701920d9d4)

要想用MAA刷隐藏坍缩范式，我们只需要让MAA在点击每一个节点后，点击确认进入节点前，查看一下坍缩范式。
若出现隐藏坍缩范式则停下，否则继续探索。
我为了偷懒只想刷第一个出现的坍缩范式，也就是一旦出现非隐藏探索范式，则退出重开。

### 检查坍缩范式

其实就是点击屏幕中上方的区域名称。

```
Sami@Roguelike@CheckCollapsalParadigms": {
    "algorithm": "OcrDetect",
    "text": [
        "初霜湖泽",
        "密静林地",
        "昧明冻土",
        "积冰岩骸",
        "无瑕花园",
        "远见之构",
        "深埋迷境"
    ],
    "cache": false,
    "roi": [
        512,
        0,
        256,
        180
    ],
    "action": "ClickSelf",
    "next": [
        "rareParadigm",
        "normalParadigm",
        "noParadigm"
    ]
}
```

### 无坍缩范式的情况

其实就是再点一下屏幕中上。
这次就偷懒不识别区域名称了，直接盲点。

```
"noParadigm": {
    "algorithm": "JustReturn",
    "action": "ClickRect",
    "specificRect": [
        635,
        5,
        10,
        10
    ],
    "next": [
        "myReturn"
    ]
},
"myReturn": {
    "algorithm": "JustReturn",
    "action": "Stop"
}
```

### 遇到稀有坍缩范式的情况

（挠头）我没搞清楚怎么直接结束主任务，而且我想之后将自定义的任务加入其他任务的sub下。
偷个懒，写个死循环。
在myStop里好好绕圈吧。

```
"rareParadigm": {
    "algorithm": "OcrDetect",
    "action": "DoNothing",
    "text": [
        "目空一些",
        "睁眼瞎",
        "图像损坏",
        "一抹黑"
    ],
    "cache": false,
    "roi": [
        0,
        300,
        640,
        360
    ],
    "next": [
        "myStop"
    ]
},
"myStop": {
    "algorithm": "JustReturn",
    "action": "DoNothing",
    "next": [
        "myStop"
    ]
}
```

### 遇到普通坍缩范式的情况

直接重开即可。

```
"normalParadigm": {
    "algorithm": "OcrDetect",
    "action": "DoNothing",
    "text": [
        "去量化",
        "去深量化",
        "实质性坍缩",
        "蔓延性坍缩",
        "非线性移动",
        "非线性行动",
        "情绪实体",
        "恐怖实体",
        "泛社会悖论",
        "泛文明悖论",
        "气压异常",
        "气压失序",
        "触发性损伤",
        "触发性危殆",
        "趋同性消耗",
        "趋同性缺失"
    ],
    "roi": [
        0,
        300,
        640,
        360
    ],
    "cache": false,
    "next": [
        "Sami@Roguelike@ExitThenAbandon"
    ]
}
```

### 将自定义任务加入现有的任务的sub

以下只是举例。我最终将`Sami@Roguelike@CheckCollapsalParadigms`加入了所有**点击事件节点**任务的sub下。
为了以防万一，将subErrorIgnored设置为true。

```
"Sami@Roguelike@StageEmergencyDps": {
    "templThreshold": 0.95,
    "sub": [ "Sami@Roguelike@CheckCollapsalParadigms" ],
    "subErrorIgnored": true
},
"Sami@Roguelike@StageEmergencyDpsAI6": {
    "templThreshold": 0.95,
    "baseTask": "Roguelike@Stage",
    "sub": [ "Sami@Roguelike@CheckCollapsalParadigms" ],
    "subErrorIgnored": true,
    "next": [
        "Sami@Roguelike@StageEmergencyDpsEnter",
        "Sami@Roguelike@StageCombatDpsEnter",
        "Sami@Roguelike@Stages#next"
    ]
}
```

### 主动漏怪

MAA的肉鸽辅助协议已经相当好用了。
每一关的战斗策略都保存在`resource/roguelike/Sami/autopilot`目录下。
第一层的战斗除了**死囚之夜**需要在特定击杀数后撤退干员，其他战斗只需要设置在击杀数1（漏一个怪）后再开始部署干员即可。
当然如果不堵门可能一下子就漏两个了，还需要根据实际情况修改战斗策略。
下面是我所使用的战斗策略（只考虑第一层）。

文件分享：
[死囚之夜.json](https://github.com/MaaAssistantArknights/MaaAssistantArknights/files/15256709/default.json)
[度假村冤魂.json](https://github.com/MaaAssistantArknights/MaaAssistantArknights/files/15256712/default.json)
[苔手.json](https://github.com/MaaAssistantArknights/MaaAssistantArknights/files/15256710/default.json)
[待宰的兽群.json](https://github.com/MaaAssistantArknights/MaaAssistantArknights/files/15256713/default.json)
[事不过四.json](https://github.com/MaaAssistantArknights/MaaAssistantArknights/files/15256705/default.json)
[半吊子之旅.json](https://github.com/MaaAssistantArknights/MaaAssistantArknights/files/15256706/default.json)

死囚之夜漏怪比较靠运气，偶尔还是会一不小心全杀了。
为了防止锏拿三技能，我还调整了`resource/roguelike/Sami/recruitment.json`

### 其他改进

为了防止MAA自动使用符文版，我清空了 `resource/roguelike/Sami/foldartal.json`。
（其实应该可以直接改`tasks.json`吧）

文件分享：
[foldartal.json](https://github.com/MaaAssistantArknights/MaaAssistantArknights/files/15256848/foldartal.json)

之后可以考虑在`Sami@Roguelike@Stages_aggressive`下把战斗的优先级调高了一些，也可以修改`encounter.json`和`shopping.json`，以尽可能快地增加坍缩值，并防止任何可能减少坍缩值的可能性。

不知是我的MBP2017跟不上时代了还是怎么，编译MAA总是遇到问题。
幸好使用maa-cli可以直接使用自己的source文件夹，曲线救国了属于是。
详情参考[MAA-CLI使用指南](https://maa.plus/docs/%E7%94%A8%E6%88%B7%E6%89%8B%E5%86%8C/user_manual_for_cli/cli-intro.html)。

碎碎念：其实我一开始是想看看代码，给MAA加点功能。可惜编译一直不成功，干脆先通过改resource文件曲线救国一下。
这与其说是简易实现方法，不如说是**简陋**实现办法……

**Comment by ABA2396** [comment]
编译是在MaaThriftController那里报错了吗？可以试试直接卸载这个项目）

**Reply by Alan-Charred** [reply]
啊编译成功了！感谢提醒！

**Comment by peter1997546** [comment]
好想知道怎么用在国际服

**Reply by Alan-Charred** [reply]
国际服改`resource/global`里对文件夹下的文件就可以了（当然那些词汇要换成对应的译名）。
我本来是想自己刷完后试试看把这个做成一个plugin加到MAA里的……可是这都4977次坍缩范式了怎么还没出第二组Orz

**Reply by Alan-Charred** [reply]
快写好了……我再加点功能就提交一个pull request

**Comment by Alan-Charred** [comment]
已提交Pull Request #9172

**Comment by KuiPer-LTW** [comment]
路过看到个非常想要的功能，请问进展如何了？

**Reply by ABA2396** [reply]
看着改的差不多了，在等 reviewe

**Reply by KuiPer-LTW** [reply]
> 看着改的差不多了，在等 reviewe

后续是会直接实装到新版本，还是需要我们手动添加？

**Reply by ABA2396** [reply]
随版本发布

**Reply by Alan-Charred** [reply]
> 路过看到个非常想要的功能，请问进展如何了？

我做完了自己想做的部分，自己在官服上用着体验良好，成功在6500+次坍缩范式后刷到了两组隐藏范式。
可能一些细节还可以优化，不过我目前没有什么好想法。
国际服适配方面我只完成了翻译，roi调整需要国际服玩家日后提供截图。

总之目前我在摆烂等review。

**Reply by KuiPer-LTW** [reply]
> > 路过看到个非常想要的功能，请问进展如何了？
> 
> 我做完了自己想做的部分，自己在官服上用着体验良好，成功在6500+次坍缩范式后刷到了两组隐藏范式。 可能一些细节还可以优化，不过我目前没有什么好想法。 国际服适配方面我只完成了翻译，roi调整需要国际服玩家日后提供截图。
> 
> 总之目前我在摆烂等review。

礼貌问一问进度，毕竟新肉鸽都快开了

**Reply by Alan-Charred** [reply]
> > > 路过看到个非常想要的功能，请问进展如何了？
> > 
> > 
> > 我做完了自己想做的部分，自己在官服上用着体验良好，成功在6500+次坍缩范式后刷到了两组隐藏范式。 可能一些细节还可以优化，不过我目前没有什么好想法。 国际服适配方面我只完成了翻译，roi调整需要国际服玩家日后提供截图。
> > 总之目前我在摆烂等review。
> 
> 礼貌问一问进度，毕竟新肉鸽都快开了

完成进度100%，审核进度未知。

**Reply by ABA2396** [reply]
> 礼貌问一问进度，毕竟新肉鸽都快开了

等这个正式版发完进下个测试版
